### PR TITLE
Fix headers_sent() with output buffering in coroutines

### PR DESCRIPTION
**Pull request by sshymko**
- Test output buffering `ob_*()` functions
- Assert invariance of `headers_sent()` state

### CODE REVIEW

**Comment by twose** [comment]
Not quite sure what the meaning of this test is?

**Comment by sshymko** [comment]
@twose 
The function `headers_sent()` should return `false` as a prerequisite for `session_*()` functions. The output buffering is the mechanism that is supposed to prevent "sending the output" along with marking "headers sent".

Currently, the failing test indicates that the issue #1828 is not fully fixed. Not sure what happened in the Swoole versions 4.0.2 -> 4.5.x, but all these functions used to work fine. Maybe I'm missing something, but I clearly had all the `session_*()` functions working alongside the PHTML template engine that uses the output buffering `ob_*()` functions.

**Comment by sshymko** [comment]
@twose 
I investigated more and turns out the behavior depends on the coroutine settings.
The test passes with `enable_coroutine = false` and fails otherwise.

Will update the tests to cover both scenarios. Will rename the tests to be about the `headers_sent()` function, not the output buffering.

**Comment by sshymko** [comment]
@twose 
Currently, the tests reveal inconsistency of the `headers_sent()` / output buffering behavior depending on whether the coroutines are enabled or disabled. Would you like me to update the coroutine test to PASS to match the current behavior? The question comes down to whether you'd like the fix to be made here or separately. The issue seems very intricate, so I'm not going to be able to develop the fix myself and will need help from your team.

```
[PASS] swoole_http_server: headers sent (coroutine disabled) [swoole_http_server/headers_sent.phpt] 
TEST 363/969              
========DIFF========
004+ headers_sent=1
004- headers_sent=0
========DONE========
[FAIL] swoole_http_server: headers sent (coroutine enabled) [swoole_http_server/headers_sent_coroutine.phpt] 
```

File: swoole_coroutine.cc [hunk 1]
```diff
@@ -525,9 +535,7 @@ void PHPCoroutine::on_close(void *arg) {
     }
 
     if (OG(handlers).elements) {
-        if (OG(active)) {
-            php_output_end_all();
```
Comment (sshymko) [inline swoole_coroutine.cc:529]: Function `php_output_end_all()` does call `OG(active)`.

**Comment by sshymko** [comment]
@twose 
Looks like I was able to fix the `headers_sent()` incompatibility with coroutines.
Could you please review the implementation and let me know whether you'd like any changes?

**Comment by twose** [comment]
I still can't understand
The functions you mentioned are all unique to FPM server mode
Coroutine and HTTP have no relation with each other
Compatible with output_buffer is reasonable as if each process has its own output target
But HTTP is only something related to your application
e.g. I can use coroutines to create many applications which have nothing to do with HTTP

For the session-related part, I did not understand at all, maybe you can provide a realistic example
(Although I think we should disable the `Session` module in swoole)

**Comment by sshymko** [comment]
@twose 
Let me try to to elaborate on the use case in more details.

You're right that the function `headers_sent()` is generally irrelevant to the Swoole environment. Applications written specifically for Swoole would use `\Swoole\Http\Request` instance (that doesn't have an equivalent method, but that's okay).

The important thing about `headers_sent()` is that it affects the built-in session functions. You're right that applications written specifically for Swoole should use a different coroutine-aware mechanism for sessions, such as [swoft-cloud/swoft-session](https://github.com/swoft-cloud/swoft-session) or [itxiao6/session](https://github.com/itxiao6/session). However, not all applications are written from scratch with considerations of the Swoole environment specifics. They can rely on the existing libraries/frameworks in the PHP ecosystem not designed with the Swoole compatibility in mind.

Some projects may attempt to retrofit an existing FastCGI/PHP-FPM application for compatibility with Swoole to achieve performance improvements by reusing bootstrapped application state in memory between requests. We have adopted [magento/magento2](https://github.com/magento/magento2) to run on Swoole to achieve [~35/45% speed up](https://twitter.com/SergiiShymko/status/1272579161924624386?s=20) for frontend and Web API requests respectively. We do disable the coroutines and successfully use the built-in PHP sessions. They work fine and allow for such projects to exist.

The problem is that there's an inconsistency in the `headers_sent()` behavior in combination with the output buffering. Everything works as expected when the coroutines are disabled: capturing the output via the `ob_*()` functions maintains the `headers_sent = 0` which satisfies preconditions of the sessions. However, the same exact code works differently with coroutines: the first request has `headers_sent = 0` but all subsequent ones have `headers_sent = 1`. There's no logical reason for such difference in the `headers_sent` flag behavior. That's why I'm proposing a simple fix for coroutines.

PLEASE, PLEASE, PLEASE do not disable the session module as that would be a show stopper for the "classical" applications, such as Magento 2, from ever running on Swoole. If anything, it's better to just decline this PR and keep the existing sessions working as they are w/o coroutines. I still think the simple `headers_sent` fix would be appropriate.

Please let me know if you have further questions. Open to discuss further offline.

**Comment by twose** [comment]
Do not worry, I understand what you're saying now, I will try to solve it in a reasonable way

**Comment by twose** [comment]
And, can you provide a simple session example to reproduce the problem you mentioned?
I am not familiar with it, I am some busy recently, It can save me some time

**Comment by sshymko** [comment]
@twose 
A basic session example came up in a very simple library (just one small class) for compatibility of the native PHP sessions with Swoole. The bug report https://github.com/upscalesoftware/swoole-session/issues/1 demonstrates how the output buffering by Twig breaks the sessions when the coroutines are enabled. However, the issue is more generic and can be reproduced with a plain output buffering w/o template engine.

Sessions are not working when coroutines are enabled:
```php
use Upscale\Swoole\Session\SessionMiddleware;

$server = new Swoole\HTTP\Server('127.0.0.1', 8080);
$server->on('request', new SessionMiddleware(function ($request, $response) {
    ob_start();
//    echo "Test";
    ob_end_clean();

    $_SESSION['data'] = $_SESSION['data'] ?? rand();
    $response->end($_SESSION['data']);
}));
$server->start();
```
```bash
curl -c cookies.txt -b cookies.txt http://127.0.0.1:8080/
215539912

curl -c cookies.txt -b cookies.txt http://127.0.0.1:8080/
PHP Warning:  session_id(): Cannot change session id when headers already sent in vendor/upscale/swoole-session/src/SessionMiddleware.php on line 62
PHP Warning:  session_start(): Cannot start session when headers already sent in vendor/upscale/swoole-session/src/SessionMiddleware.php on line 63
PHP Warning:  session_id(): Cannot change session id when headers already sent in vendor/upscale/swoole-session/src/SessionMiddleware.php on line 80
```

The same implementation works fine when coroutines are disabled:
```php
use Upscale\Swoole\Session\SessionMiddleware;

$server = new Swoole\HTTP\Server('127.0.0.1', 8080);
$server->set([
    'enable_coroutine' => false,
]);
$server->on('request', new SessionMiddleware(function ($request, $response) {
    ob_start();
//    echo "Test";
    ob_end_clean();

    $_SESSION['data'] = $_SESSION['data'] ?? rand();
    $response->end($_SESSION['data']);
}));
$server->start();
```
```bash
curl -c cookies.txt -b cookies.txt http://127.0.0.1:8080/
2031979775

curl -c cookies.txt -b cookies.txt http://127.0.0.1:8080/
2031979775
```

**Pull request closed without merging**